### ISSUE TITLE
Acknowledge Purchase does not work for Android

### ISSUE THREAD

**Issue by hakkikonu** [dooboolab/react-native-iap#1006]
I try to Acknowledge **consumable** products but it is not working. I'm getting `{"acknowledged":false}` response with receipt

I tried two methods according to docs.

v:4.4.8
platform: android
device: real device

```
try {
          if (Platform.OS === 'ios') {
            RNIap.finishTransactionIOS(purchase.transactionId);
          } else if (Platform.OS === 'android') {
            // If consumable (can be purchased again)
            RNIap.consumePurchaseAndroid(purchase.purchaseToken);
          }
        } catch (ackErr) {
          console.warn('ackErr', ackErr);
        }
```

and

```
 try {
          const ackResult = await finishTransaction(purchase, true);
          console.log('ackResult', ackResult);
        } catch (ackErr) {
          console.warn('ackErr', ackErr);
        }
```

Errors:

```
Possible Unhandled Promise Rejection (id: 1):
Error: Google is indicating that we have some issue connecting to payment.
Error: Google is indicating that we have some issue connecting to payment.
```

But I can see purchase in my Google Play console and its accepted.

I'm testing my purchases with Google Test Account, and Android docs says:

https://developer.android.com/google/play/billing/billing_library_overview#test_acknowledging_purchase_with_license_testers

**Test acknowledging purchase with license testers**
_For purchases made by license testers, the acknowledgement window is shorter. Instead of three days, purchases are refunded and revoked if they are not acknowledged within **five minutes**._

After 5 minutes my purchases there (actually I'm trying for 3-4 hours and all purchases there). There is no problem on Google Play side but some things are not consistent.

`{"acknowledged":false}` worries me

**Comment by hyochan** [comment]
Could someone else share anything about this? I've not experienced this yet.

**Comment by jsfit** [comment]
Getting same error 
```
Possible Unhandled Promise Rejection (id: 0):
Error: Google is indicating that we have some issue connecting to payment.
```
catch block executed most of the time

```
    purchaseUpdateSubscription = purchaseUpdatedListener(
      async (purchase: InAppPurchase | SubscriptionPurchase) => {
        const receipt = purchase.transactionReceipt;
        if (receipt) {
          try {
            const ackResult = await finishTransaction(purchase);
            console.warn('ackResult', ackResult);
          } catch (ackErr) {
             console.warn('ackErr', ackErr);
          }
```

subscription is done
because getting purchases response
```
const purchases = await RNIap.getAvailablePurchases()
```

**Comment by triplumix** [comment]
I'm experiencing the same issue. Purchase is done, I can see it in Publisher Console, but then it gets refunded.

**Comment by ryandoss** [comment]
Has anyone found a solution for this? Have some very frustrated clients...

**Comment by Decoder3-14** [comment]
> Has anyone found a solution for this? Have some very frustrated clients...

Have you found the mystery behind this?

**Comment by stefanoromanello** [comment]
Any update on this issue?

**Comment by yusufpamukcu** [comment]
+++++

**Comment by wdospinal** [comment]
Hi, I'm getting the same problem and also no clue about it

Sentry is reporting:
`"ackErr",{"code":"E_DEVELOPER_ERROR","name":"Error","column":1111,"nativeStackAndroid":"[Array]","stack":"index.android.bundle:21:1111\nindex.android.bundle:1441:12207\ns@index.android.bundle:740:3079\nindex.android.bundle:740:2380\nindex.android.bundle:740:2070\nc@index.android.bundle:129:205\nX@index.android.bundle:129:1586\n_@index.android.bundle:129:484\nl@index.android.bundle:740:1833\nandroid@index.android.bundle:1441:11967\nF@index.android.bundle:1441:12338\nindex.android.bundle:1440:3666\ny@index.android.bundle:113:587\nindex.android.bundle:113:1890\ny@index.android.bundle:113:587\no@index.android.bundle:113:1066\nindex.android.bundle:113:1209\nf@index.android.bundle:129:155\nindex.android.bundle:129:864\ny@index.android.bundle:134:581\nC@index.android.bundle:134:947\ncallImmediates@index.android.bundle:134:3004\nvalue@index.android.bundle:30:2919\nindex.android.bundle:30:1095\nvalue@index.android.bundle:30:2611\nvalue@index.android.bundle:30:1065\nvalue@[native code]\nvalue@[native code]","sourceURL":"index.android.bundle","userInfo":null,"line":21,"message":"Google is indicating that we have some issue connecting to payment."}]`

**Comment by yusufpamukcu** [comment]
> I'm experiencing the same issue. Purchase is done, I can see it in Publisher Console, but then it gets refunded.

Hi, Although the purchase was completed, it was making a refund and I solved it with the following function.

requestPurchase = async () => {
    await RNIap.initConnection();
    try {
      await RNIap.requestPurchase("productCode", true);
       await RNIap.purchaseUpdatedListener(
        purchase => {
          RNIap.consumePurchaseAndroid(purchase.purchaseToken);
          RNIap.acknowledgePurchaseAndroid(purchase.purchaseToken);
        // the purchase has been made
        });
    } catch (err) {
      console.log(err)
       console.warn(err.code, err.message);
    }
  }

**Comment by andersmurphy** [comment]
Any updates on this? Billing is basically broken for android if you use `finishTransaction`.

**Comment by StefanoCremona** [comment]
I had the same problem.
Using react-native-iap, 5.0.1 docs say I have just to use finishTransaction, that fails each time with the error reported and cancelling the subscriptions after a while.

I solved it with a try catch statement but the weird thing is that using the following code, I see all the logs logs either the promise is resolved or rejected:

`RNIap.finishTransaction(purchase)
                    .then(resp =>
                      console.log(
                        'purchaseUpdatedListener: finishTransaction',
                        resp,
                      ),
                    )
                    .catch(err => {
                      console.warn(
                        'purchaseUpdatedListener: finishTransaction err ',
                        err.code,
                        err.message,
                      )
                    })`
and these are the logs:
`purchaseUpdatedListener: finishTransaction {message: "", code: "OK", debugMessage: "", responseCode: 0}
purchaseUpdatedListener: finishTransaction err  E_DEVELOPER_ERROR Google is indicating that we have some issue connecting to payment.`

**Comment by StefanoCremona** [comment]
I had the same problem trying to purchase a subscription.
This product can't be considered a consumable, neither a non consumable one, so I just removed the second parameter in the finishTransaction call:

RNIap.finishTransaction(purchase)

and it works.

If a put back true, I have the behaviour you have documented @hakkikonu

**Comment by freezen82** [comment]
I saw codes somewhere, maybe stack overflow, that code do get purchase data from google play after finishTrasaction().
its acknowledged is true.
My code modified from original, but works fine.

`const checkReceiptAndroid = async (currentPurchase) => {
    let validatedReceipt = null;
    const receipt = currentPurchase.purchaseToken;
    if (receipt) {
        try {
            const purchases = await getAvailablePurchases();
            if (purchases.length) {
                const currentPurchase = purchases.find((v) => v.purchaseToken === receipt);
                validatedReceipt = currentPurchase;
            }
        } catch (error) {
            validatedReceipt = false;
        }
    }
    return validatedReceipt;
};

await RNIap.finishTransaction(purchase);
const validatedReceipt = await checkReceipt(purchase, isDev);`

result: 
`{"autoRenewingAndroid": true, "developerPayloadAndroid": "", "isAcknowledgedAndroid": true, "obfuscatedAccountIdAndroid": "", "obfuscatedProfileIdAndroid": "", "orderId": "GPA.3330-0770-5858-12862", "packageNameAndroid": "com.mlab.aveshop", "productId": "com.mlab.aveshop.subsc.1month", "purchaseStateAndroid": 1, "purchaseToken": "lghcffcjdmfimgdcipnaklnh.AO-J1OwdC7FOIBYE7XjSQAEbr4TeohY97PHN2TFjfqMq25reqTyBLTb7Z8GV-kkBT1EaGjnsnY78VMJZvDwhMXLq-yP7kXyAZg", "signatureAndroid": "I5KcJjv98lKf5cgLVGLdWtKpW+5+Hxd/BwWK3OF7zjL5FbhW1G8/cstVTYGN1Dad9sTcs3bqElCZciUWRKrYA+2vMqqKWcRkQZsv8x48lGmU0th2D48sw7U4Lth7bpQ5IVlBWihrLy6J6hrE4s5fbcAhXKWAuIfcKUDXui0KzOIygbRNN0hJm8Ptnhk2MiahXxld9IXju+wVTYTUCvWkIKESVA1nxm/G+eot7XtxXV/lV0Y+KomchXaA2s4hF8gDCUY2DtT/jWHq5eKx+C/IGoy5VT1MTN2M+BGIOHF6pRjguCadnYZQzQXrTJRFneaDN3jGIlsxkO5ES7A6+4GKjA==", "transactionDate": 1625167624189, "transactionId": "GPA.3330-0770-5858-12862", "transactionReceipt": "{\"orderId\":\"GPA.3330-0770-5858-12862\",\"packageName\":\"com.mlab.aveshop\",\"productId\":\"com.mlab.aveshop.subsc.1month\",\"purchaseTime\":1625167624189,\"purchaseState\":0,\"purchaseToken\":\"lghcffcjdmfimgdcipnaklnh.AO-J1OwdC7FOIBYE7XjSQAEbr4TeohY97PHN2TFjfqMq25reqTyBLTb7Z8GV-kkBT1EaGjnsnY78VMJZvDwhMXLq-yP7kXyAZg\",\"autoRenewing\":true,\"acknowledged\":true}"}`

**Comment by andresesfm** [comment]
This has been solved in 7.3.0. `finishTransaction` will call `consumePurchaseAndroid or `acknowledgePurchaseAndroid` depending on the parameter second parameter that indicates if the purchase was for a consumable.

This thread is a bit old and doesn't reflect the newest changes. If the issue remains on `7.3.0` or newer I'll be happy to take a look. Closing for now, please create a new issue if needed
